### PR TITLE
[ISSUE-106] Attachment should accept a ReadStream

### PR DESCRIPTION
**Pull request by gaol**
Issue: #106

This is the PR target master branch

### CODE REVIEW

**Comment by vietj** [comment]
I see currently that `MailAttachment` is a data object, I believe that to use `ReadStream` it should be converted to be rather an `@VertxGen` in Vert.x 4

**Comment by vietj** [comment]
@pmlopes WDYT ?

**Comment by gaol** [comment]
@vietj  May I know the reason to convert the data object to use `@VertxGen` ? :)

**Comment by vietj** [comment]
so it can be used in polyglot languages of vertx

**Comment by gaol** [comment]
I am trying to convert to use `@VertxGen` in branch: https://github.com/gaol/vertx-mail-client/commits/ehencement/issue-106-master-vertxgen

It seems will cause 2 problems:
- The generated `*Converter.java` are missing from `src/main/generated` directory, not only for the `MailAttachment`one
- The `MailAttachment` in [dataobjects](https://github.com/vert-x3/vertx-mail-client/blob/master/src/main/asciidoc/dataobjects.adoc#mailattachment) gets removed as well.

@vietj  @pmlopes Do you have any suggestions on that? Thank you :)

**Comment by vietj** [comment]
normally it's fine and expected to have those removed, the converter should not be needed anymore

**Comment by vietj** [comment]
this looks like an improvement. I think we need to document the breaking changes for Vert.x 4

**Comment by gaol** [comment]
What is the process to update wiki page ? https://github.com/vert-x3/wiki/wiki/4.0.0-Deprecations-and-breaking-changes in this case

**Comment by vietj** [comment]
I think only vertx commiters can do it, let me check

**Comment by vietj** [comment]
can we actually do this in two separate PRs ? one that changes data object to vertx gen and then another one that add attachment support

**Comment by gaol** [comment]
@vietj  Sure, I can do that. Can you create an issue for the data object to vertx gen changes, I can make a PR for that first. :)

**Comment by vietj** [comment]
@gaol here it is https://github.com/vert-x3/vertx-mail-client/issues/109

**Comment by gaol** [comment]
Please on hold this PR, waiting for #110 first. :)

**Comment by vietj** [comment]
@gaol can you rebase this PR on master ?

**Comment by gaol** [comment]
@vietj Sorry for late update, I was busy on other issues. :) Please review again, thanks :)

**Comment by vietj** [comment]
thanks @gaol

**Comment by gaol** [comment]
@pmlopes would you please review? Thanks :)

**Comment by vietj** [comment]
I've started to review this PR.

I think that we should simplify some code already existing now in vertx-core for piping a ReadStream to a WriteStream.

So we should:

- decorate the body stream  used by `SMTPSendMail#writeStream` so that it does the job in `SMTPSendMail#writeStream`
- use this new decorated stream to pipe it to the `NetSocket` which is a `WriteStream<Buffer>` using the `ReadStream.pipe` operation
- when using the `ReadStream#pipe` operation the `Pipe` object should be configured to avoid closing the `WriteStream<Buffer>` on completion

This will implement the back-pressure between the `ReadStream` and the `WriteStream` and also it will handle the error management when the `ReadStream` exception handler reports an exception and we can gracefully handle the error when sending the mail.

**Comment by gaol** [comment]
@vietj Sorry, I am struggling on the DKIM issue currently, which should be ready soon, I will try to get back to this issue very soon. :)

**Comment by gaol** [comment]
Updated to introduce a `BodyReadStream` in `SMTPSendMail` to pipe to the underline NetSocket.

@vietj Would you please review again?  many thanks :)

**Comment by vietj** [comment]
sure @gaol

File: src/test/java/io/vertx/ext/mail/TestUtils.java [hunk 1]
```diff
@@ -69,4 +66,15 @@ public static String inputStreamToString(final InputStream inputStream) throws I
     return string.toString();
   }
 
+  // this method can be deprecated after JDK9+ which has InputStream.readAllBytes()
+  public static byte[] inputStreamToBytes(final InputStream inputStream) throws IOException {
+    ByteArrayOutputStream bufferedArray = new ByteArrayOutputStream();
+    int nRead;
+    byte[] buffer = new byte[4096];
```
Comment (vietj) [inline src/test/java/io/vertx/ext/mail/TestUtils.java]: this seems a bit large, 512 would be better

Comment (gaol) [reply src/test/java/io/vertx/ext/mail/TestUtils.java]: Done by changing to 512

**Comment by gaol** [comment]
Squashed the 2nd commit of updating `TestUtils` into 1.

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 2]
```diff
@@ -40,17 +48,18 @@
   private final MailConfig config;
   private final Handler<AsyncResult<MailResult>> resultHandler;
   private final MailResult mailResult;
-  private final String hostname;
-
-  private String mailMessage;
+  private final EncodedPart encodedPart;
+  private final AtomicLong wrottern = new AtomicLong();
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: I think this is a typo instead if should be `written`

File: src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java [hunk 3]
```diff
@@ -137,30 +137,14 @@ void writeLine(String str, boolean mayLog) {
...
     ns.write(str + "\r\n");
-  }
-
-  // write single line not expecting a reply, using drain handler
-  void writeLineWithDrainHandler(String str, boolean mayLog, Handler<Void> handler) {
-    if (mayLog) {
-      log.debug(str);
-    }
     if (ns.writeQueueFull()) {
+      ns.pause();
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java]: why is this needed ? this will pause reading from the same netsocket

File: src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java [hunk 4]
```diff
@@ -137,30 +137,14 @@ void writeLine(String str, boolean mayLog) {
...
       ns.drainHandler(v -> {
-        // avoid getting confused by being called twice
-        ns.drainHandler(null);
-        ns.write(str + "\r\n");
-        handler.handle(null);
+        ns.resume();
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java]: why is this needed ? this will resume reading from the same netsocket

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 5]
```diff
@@ -187,57 +202,159 @@ private void dataCmd() {
...
+      sendMailHeaders(part);
+      sendMultiPart(part, 0, endHandler);
+    } else {
+      sendRegularPart(part, endHandler);
+    }
+  }
+
+  private void sendMultiPart(EncodedPart part, int i, Handler<Void> endHandler) {
+    // write boundary start
+    final StringBuilder sb = new StringBuilder("--").append(part.boundary());
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: no need for to use a `StringBuilder`, a simple string works

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 6]
```diff
@@ -187,57 +202,159 @@ private void dataCmd() {
...
+
+  private Handler<Void> handlerInContext(Handler<Void> handler) {
+    return vv -> connection.getContext().runOnContext(v -> handler.handle(null));
+  }
+
+  // what we need: strings line by line with CRLF as line terminator
+  private class BodyReadStream implements ReadStream<Buffer> {
+
+    private final ReadStream<Buffer> stream;
+    private final int size = 57;
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: what is this 57 magic number ?

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 7]
```diff
@@ -187,57 +202,159 @@ private void dataCmd() {
...
+      if (handler == null) {
+        return this;
+      }
+      this.handler = handler;
+      stream.handler(b -> connection.getContext().runOnContext(v -> {
+        Buffer buffer = streamBuffer.appendBuffer(b);
+        int start = 0;
+        while(start + size < buffer.length()) {
+          final String theLine = Utils.base64(buffer.getBytes(start, start + size));
+          handler.handle(Buffer.buffer(theLine + "\r\n"));
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: we need to call the `handler` only a single time because this stream implementation delegates the `fetch` operation and does not calculate the number of fetches. For instance if this stream is fetched `1` but the provided `Buffer` is actually split in `4` then the handler will be called `4` times although it only asked for `1` buffer.

File: src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java (hunk 3 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java]: This must be misunderstanding of how the Streams are used. :)

File: src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java (hunk 4 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPConnection.java]: This must be misunderstanding of how the Streams are used. :)

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java (hunk 6 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: 57 / 3 * 4 = 76. 76 + "\r\n".length() = 78. which is the most common used length for email attachment. See: https://tools.ietf.org/html/rfc5322#section-2.1.1

Use this size for each byte[] --> base64String conversion to make sure to get limited length string.

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java (hunk 7 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: Ah, that is the thing I missed, thank you. I will fix it.

Comment (vietj) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: yes this is a bit tricky

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java (hunk 6 above)
Comment (vietj) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: can you add a comment explaining this ?

**Comment by gaol** [comment]
Updated with the following changes:
* Fix typo: `wrottern` --> `written`
* Reset `SMTPConnection` to use `writeLineWithDrainHandler` instead for the back-pressure support
* Use simple string instead of `StringBuilder`
* Add a comment on the magic number: `57` for MailAttachement ReadStream
* Compose the buffer in the loop to manipulate multiple `78` long strings buffer to handle once outside of the loop.

@vietj Would you please review again? Thanks :)

File: src/main/java/io/vertx/ext/mail/MailAttachment.java [hunk 8]
```diff
@@ -74,6 +75,41 @@ static MailAttachment create(MailAttachment other) {
...
+   */
+  @Fluent
+  MailAttachment setStream(ReadStream<Buffer> stream);
+
+  /**
+   * Gets the size of the attachment.
+   *
+   * @return the size of the attachment
+   */
+  Integer getSize();
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/mail/MailAttachment.java]: Why not a primitive? `int`

File: src/main/java/io/vertx/ext/mail/MailAttachment.java [hunk 9]
```diff
@@ -74,6 +75,41 @@ static MailAttachment create(MailAttachment other) {
...
+   * Sets the size of the attachment.
+   *<p>
+   * It is needed when using ReadStream for the MailAttachement.
+   *</p>
+   *
+   * @param size the size of the attachment
+   * @return this to be able to use it fluently
+   */
+  @Fluent
+  MailAttachment setSize(Integer size);
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/mail/MailAttachment.java]: Same here, why not just `int`?

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 10]
```diff
@@ -187,57 +202,177 @@ private void dataCmd() {
   }
 
   private void sendMaildata() {
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: Shouldn't this method become async? it calls an async method so if you're not on the right timing you might get the state wrong or be unable to catch any eventual error.

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java [hunk 11]
```diff
@@ -187,57 +202,177 @@ private void dataCmd() {
...
-  /**
-   * create message if it hasn't been already
-   */
-  private void createMailMessage() {
-    if (mailMessage == null) {
-      MailEncoder encoder = new MailEncoder(email, hostname);
-      mailMessage = encoder.encode();
-      mailResult.setMessageID(encoder.getMessageID());
+  private Handler<Void> handlerInContext(Handler<Void> handler) {
+    return vv -> connection.getContext().runOnContext(v -> handler.handle(null));
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: i think
```java
return vv -> connection.getContext().runOnContext(handler);
```

Would have the same result and it avoid 1 extra lambda

**Comment by vietj** [comment]
the PR should be rebased as now there are conflicting changes with the futurisation work

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java (hunk 11 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: :+1:

File: src/main/java/io/vertx/ext/mail/MailAttachment.java (hunk 9 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/MailAttachment.java]: I was thinking to check whether the size if set or not. But looks like using `0` as not-set works well too. :)

Comment (vietj) [reply src/main/java/io/vertx/ext/mail/MailAttachment.java]: `-1` would be better I think

File: src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java (hunk 10 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java]: What I understand on this method is that it will be called only once after getting a succeed result of writing `DATA`  in method: `dataCmd()`.

I need to make some modification to make this async.

**Comment by gaol** [comment]
Thanks all for the reviews. :)

Updated this time includes:
* Update MailAttachment.size from `Integer` to `int`
* Update `SMTPSendMail.handlerInContext` method to save 1 lambada
* Refactor `SMTPConnection.writeLineWithDrainHandler` to `SMTPConnection.writeLineWithDrainPromise` with last parameter as a `Promise<Void>`
* Use `Promise<Void>` instead of `Handler<Void>` in `SMTPSendMail` methods to be able to propagate the eventual exceptions, including `SMTPSendMail.sendMaildata`

**Comment by gaol** [comment]
@vietj @pmlopes Would you please review again? Thanks :)

**Comment by pmlopes** [comment]
It looks fine by me, waiting on @vietj feedback.

**Comment by vietj** [comment]
merged thanks for the contribution
